File: solutions/kotlin/sum-of-multiples/1/src/main/kotlin/SumOfMultiples.kt
object SumOfMultiples {

    fun sum(factors: Set<Int>, limit: Int): Int {
        var numbers = mutableSetOf<Int>()
        var sum = 0
        for(f in factors) {
            if(f < 1 || f >= limit) continue
            val upper = if(limit % f == 0) limit / f - 1 else limit / f
            for(m in 1..upper) {
                val num = m * f
                if(numbers.add(num)) sum += num
            }
        }

        return sum
    }
}

File: solutions/zig/sum-of-multiples/1/sum_of_multiples.zig
const std = @import("std");
const mem = std.mem;

pub fn sum(allocator: mem.Allocator, factors: []const u32, limit: u32) !u64 {
    _ = allocator;
    var total: u64 = 0;

    for (1..limit) |m| {
        for (factors) |f| {
            if (f > 0 and m % f == 0) {
                total += m;
                break;
            }
        }
    }

    return total;
}
